File: 프로그래머스/2/12905. 가장 큰 정사각형 찾기/가장 큰 정사각형 찾기.js
function solution(board) {
  const row = board.length;
  const col = board[0].length;

  // 각 칸까지 끝나는 가장 큰 정사각형 한 변의 길이 저장
  const dp = Array.from({ length: row }, () => Array(col).fill(0));
  let maxSize = 0;

  for (let i = 0; i < row; i++) {
    for (let j = 0; j < col; j++) {
      if (board[i][j] === 0) {
        dp[i][j] = 0;                 // 0이면 정사각형 불가
      } else if (i === 0 || j === 0) {
        dp[i][j] = 1;                 // 첫 행/열에서 1이면 길이 1 정사각형
      } else {
        dp[i][j] = Math.min(
          dp[i - 1][j],               // 위
          dp[i][j - 1],               // 왼
          dp[i - 1][j - 1]            // 왼위
        ) + 1;
      }

      maxSize = Math.max(maxSize, dp[i][j]);
    }
  }

  return maxSize * maxSize;
}

File: 프로그래머스/2/12905. 가장 큰 정사각형 찾기/가장 큰 정사각형 찾기.py
def solution(board):
    x_len = len(board[0])
    y_len = len(board)
    
    dp = [[0] * x_len for _ in range(y_len)]
    maxSize = 0;
    
    for y in range(y_len):
        for x in range(x_len):
            if board[y][x] == 1:
                if x == 0 or y == 0:
                    dp[y][x] = 1 # 첫행이나 첫열은 무조건 1
                else:
                    dp[y][x] = min(dp[y-1][x], dp[y][x-1], dp[y-1][x-1]) + 1
                
                maxSize = max(maxSize, dp[y][x])
            
    return maxSize ** 2